File: 02-longest-subarray-sum-k/solution.go
// Longest Subarray with Sum K
// Time: O(n), Space: O(n)

package main

import "fmt"

func longestSubarrayWithSumK(arr []int, k int) int {
    prefixSumMap := make(map[int]int)
    sum := 0
    maxLen := 0
    
    for i, num := range arr {
        sum += num
        
        if sum == k {
            maxLen = i + 1
        }
        
        if idx, found := prefixSumMap[sum - k]; found {
            if i - idx > maxLen {
                maxLen = i - idx
            }
        }
        
        if _, found := prefixSumMap[sum]; !found {
            prefixSumMap[sum] = i
        }
    }
    
    return maxLen
}

// Test
func main() {
    result := longestSubarrayWithSumK([]int{1, 2, 3, 1, 1, 1, 1}, 3)
    fmt.Println(result) // Output: 3
}

File: 02-longest-subarray-sum-k/solution.cpp
// Longest Subarray with Sum K
// Time: O(n), Space: O(n)

#include <vector>
#include <unordered_map>
#include <iostream>
#include <algorithm>

using namespace std;

class Solution {
public:
    int longestSubarrayWithSumK(vector<int>& arr, int k) {
        unordered_map<int, int> prefixSumMap;
        int sum = 0;
        int maxLen = 0;
        
        for (int i = 0; i < arr.size(); i++) {
            sum += arr[i];
            
            if (sum == k) {
                maxLen = i + 1;
            }
            
            if (prefixSumMap.find(sum - k) != prefixSumMap.end()) {
                maxLen = max(maxLen, i - prefixSumMap[sum - k]);
            }
            
            if (prefixSumMap.find(sum) == prefixSumMap.end()) {
                prefixSumMap[sum] = i;
            }
        }
        
        return maxLen;
    }
};

// Test
int main() {
    Solution sol;
    vector<int> arr = {1, 2, 3, 1, 1, 1, 1};
    cout << sol.longestSubarrayWithSumK(arr, 3) << endl; // Output: 3
    return 0;
}
